pkg/daemon/daemon.go

```diff
@@ -1276,17 +1277,33 @@ func (dn *Daemon) validateOnDiskState(currentConfig *mcfgv1.MachineConfig) error
 		return errors.Errorf("expected target osImageURL %s", currentConfig.Spec.OSImageURL)
 	}
 	// And the rest of the disk state
-	currentIgnConfig, err := ctrlcommon.ParseAndConvertConfig(currentConfig.Spec.Config.Raw)
+	// We want to verify the disk state in the spec version that it was created with,
+	// to remove possibilities of behaviour changes due to translation
+	ignconfigi, err := ctrlcommon.IgnParseWrapper(currentConfig.Spec.Config.Raw)
 	if err != nil {
 		return errors.Errorf("Failed to parse Ignition for validation: %s", err)
 	}
-	if err := checkFiles(currentIgnConfig.Storage.Files); err != nil {
-		return err
-	}
-	if err := checkUnits(currentIgnConfig.Systemd.Units); err != nil {
-		return err
+
+	switch typedConfig := ignconfigi.(type) {
+	case ign3types.Config:
```

Comment:
for validation I'm thinking we should keep validating in the config that it was created in

Reply:
doesnt the ign translator func also validate?

Reply:
Right in this case I meant validating the on-disk config based on the rendered-config

If the cluster is still on a spec 2 rendered-config version (which shouldn't happen that often, only should happen if a) templates changed causing a race or b) a rendered-config was created during a middle of an update or c) a previous rendered-config update was failing), I'm thinking we should continue to validate in spec 2, instead of translating up and then validating in spec 3.

Maybe we could get away with translating first, just have to be a bit more careful